### ISSUE TITLE
Updates not persisting on cordova-ios@4.x

### ISSUE THREAD

**Issue by philly-d** [driftyco/ionic-plugin-deploy#60]
Ionic Deploy, like the rest of the Ionic platform, is awesome! But I'm having issues adding it to an existing project running cordova-ios @4.x (I tried with 4.0.1 and 4.1.0).
Deploy works great when I run through the tutorial here: http://docs.ionic.io/docs/deploy-from-scratch.
This uses the Ionic tabs starter app, which, via the Ionic CLI, currently adds the iOS platform using cordova-ios @3.8.
On an existing project @4.x, deploying an update still works great -- the app gets the update and installs it -- but when I quit + restart the app, the changes don't persist, and deploy says there are no updates available. (Strangely, running deploy.load() at this point loads the update, but then restarting the app reverts the changes.)
I ran into this bug because Ionic Push and the push notification plugin phonegap-plugin-push now require Cordova-ios @4.x.

Here's my platform info:
> ionic plugin list
cordova-plugin-console 1.0.2 "Console"
cordova-plugin-device 1.1.1 "Device"
cordova-plugin-splashscreen 3.1.0 "Splashscreen"
cordova-plugin-statusbar 2.1.2 "StatusBar"
cordova-plugin-whitelist 1.2.1 "Whitelist"
ionic-plugin-deploy 0.5.0 "IonicDeploy"
ionic-plugin-keyboard 2.0.1 "Keyboard"

> ionic info
Your system information:

Cordova CLI: 6.0.0
Gulp version:  CLI version 3.8.8
Gulp local:   Local version undefined
Ionic Version: 1.2.4-nightly-1917
Ionic CLI Version: 1.7.14
Ionic App Lib Version: 0.7.0
ios-deploy version: 1.8.5
ios-sim version: 5.0.8
OS: Mac OS X El Capitan
Node Version: v5.0.0
Xcode version: Xcode 7.3 Build version 7D175

It's probably easiest to reproduce by following the Deploy tutorial (http://docs.ionic.io/docs/deploy-from-scratch), but running `ionic platform remove ios` and `ionic platform add ios@4.1.0` before running the app on the emulator and deploying an update.

**Comment by philly-d** [comment]
I did some debugging, and I think this is a bug in Cordova-ios, so I'm going to try to open an issue with them when I get a chance. I'm sorry for any inconvenience here if anyone's already been looking into this.

In the meantime, for anyone else having the same issue, what I found is that the Cordova method `onReset` isn't getting called when the webview starts loading, which is what Ionic Deploy overrides to know when to redirect the webview to the new downloaded Deploy directory.  I think I've got 2 workarounds.

One is to try to fix the Cordova issue, which seems to be that the notification it posts to trigger the `onReset` method is being assigned to a nil pointer, while the subscriber is listening for it on the web view instance.
Here is where the notification is being subscribed to: https://github.com/apache/cordova-ios/blob/f75bf67438257132e739238ed579d73271b3a716/CordovaLib/Classes/Public/CDVPlugin.m#L69
And here is where it's being posted:  https://github.com/apache/cordova-ios/blob/f75bf67438257132e739238ed579d73271b3a716/CordovaLib/Classes/Private/Plugins/CDVUIWebViewEngine/CDVUIWebViewNavigationDelegate.m#L48
In the latter, you can change line 48 to:
`[[NSNotificationCenter defaultCenter] postNotification:[NSNotification notificationWithName:CDVPluginResetNotification object:theWebView]];`

The other option in the meantime is to add the new Cordova-ios method `shouldOverrideLoadWithRequest` to IonicDeploy.m and call `doRedirect` when the webview is first being loaded. Here's the method from the Cordova release notes: https://github.com/apache/cordova-ios/blob/master/guides/API%20changes%20in%204.0.md#optional

**Comment by ericb** [comment]
Thanks for the update @philly-d

**Comment by NicksonYap** [comment]
@philly-d I am having the same problem here, update is not persistent only on iOS, persists on Android.

I downloaded the zip from master (apparently version: 4.2.0-dev) and changed line 48 of the file `CDVUIWebViewNavigationDelegate.m` 

ran command:
```
ionic platform rm ios
```
then 
```
cordova platform add --link /Users/nickson/Desktop/cordova-ios
```
Adding ios project...
iOS project created with cordova-ios@4.2.0-dev and has a linked CordovaLib
Installing "cordova-plugin-console" for ios
Installing "cordova-plugin-device" for ios
Installing "cordova-plugin-splashscreen" for ios
Installing "cordova-plugin-statusbar" for ios
Installing "cordova-plugin-whitelist" for ios
Installing "ionic-plugin-deploy" for ios
Installing "ionic-plugin-keyboard" for ios
Running command: /Users/nickson/desktop/testiOSdeploy/hooks/after_prepare/010_add_platform_class.js /Users/nickson/desktop/testiOSdeploy
add to body class: platform-ios
```

so i did the usual `ionic run ios` then test the deploy, same effect.

Were you posting a solution or just an untested potential method for a work-around?

**Comment by NicksonYap** [comment]
Summary of persistence test on cordova-ios versions based on a fresh ionic project (tabs) on iOS 8.0 and iOS 9.3 (simulator):

Not persistent:
4.2.0-dev (original from GitHub not persistent)
4.1.1
4.0.0

Persistent:
4.2.0-dev (after applying workaround by @philly-d)
3.9.2
3.8.0

I will stick with 3.9.2 even if it's depreciated, because Im uncertain if the workaround will cause other problems

**Basically, doing this will fix the persistence issue due to cordova-ios:**
```
ionic platform rm ios
ionic platform add ios@3.9.2
```
you might need to attempt building/running twice for build to succeed.


My environment:

`ionic info`
```
Your system information:

Cordova CLI: 6.1.1
Ionic Version: 1.2.4
Ionic CLI Version: 1.7.14
Ionic App Lib Version: 0.7.0
ios-deploy version: 1.8.6 
ios-sim version: 5.0.8 
OS: Mac OS X El Capitan
Node Version: v5.10.1
Xcode version: Xcode 7.3 Build version 7D175 
```

`ionic plugin list`
```
cordova-plugin-console 1.0.2 "Console"
cordova-plugin-device 1.1.1 "Device"
cordova-plugin-splashscreen 3.2.1 "Splashscreen"
cordova-plugin-statusbar 2.1.2 "StatusBar"
cordova-plugin-whitelist 1.2.1 "Whitelist"
ionic-plugin-deploy 0.5.0 "IonicDeploy"
ionic-plugin-keyboard 2.0.1 "Keyboard"
```

**Comment by Fuiste** [comment]
Should be fixed as of https://github.com/driftyco/ionic-plugin-deploy/releases/tag/v0.5.1

Let me know if the issue persists.

**Comment by asafco4** [comment]
Hey @Fuiste , I have version 0.5.0, what's the best way to upgrade it to 0.5.1?
I tried to remove the plugin and install it but it's also 0.5.0 again...

**Comment by adinu** [comment]
hi,

we tried to update to 0.5.2/0.5.1 version (Because of the problem with the IOS) and for some reason the application doesn't open.

we have tried updating it with "ionic state reset" and also with "remove plugin" but the issue occurs.

this is the log we have on IOS:
2016-04-25 11:58:57.880 OurCrowd[512:139471] Apache Cordova native platform version 4.1.1 is starting.
2016-04-25 11:58:57.881 OurCrowd[512:139471] Multi-tasking -> Device: YES, App: YES
2016-04-25 11:58:57.963 OurCrowd[512:139471] Using UIWebView
2016-04-25 11:58:57.965 OurCrowd[512:139471] [CDVTimer][handleopenurl] 0.065982ms
2016-04-25 11:58:57.966 OurCrowd[512:139471] Unlimited access to network resources
2016-04-25 11:58:57.966 OurCrowd[512:139471] [CDVTimer][intentandnavigationfilter] 1.505971ms
2016-04-25 11:58:57.966 OurCrowd[512:139471] [CDVTimer][gesturehandler] 0.043988ms
2016-04-25 11:58:57.978 OurCrowd[512:139471] [CDVTimer][splashscreen] 11.386991ms
2016-04-25 11:58:57.985 OurCrowd[512:139471] [CDVTimer][statusbar] 6.640971ms
2016-04-25 11:58:57.985 OurCrowd[512:139471] [CDVTimer][keyboard] 0.503004ms
2016-04-25 11:58:57.986 OurCrowd[512:139471] VERSION LABEL: 1.4.6:1461574633:f7048708-0a19-11e6-8393-6a2cae197324
2016-04-25 11:58:57.986 OurCrowd[512:139471] updating version label
2016-04-25 11:58:57.987 OurCrowd[512:139471] [CDVTimer][ionicdeploy] 1.514018ms
2016-04-25 11:58:57.987 OurCrowd[512:139471] [CDVTimer][TotalPluginStartup] 22.298992ms
2016-04-25 11:58:57.991 OurCrowd[512:139471] createNotificationChecker
2016-04-25 11:58:57.991 OurCrowd[512:139471] not coldstart
2016-04-25 11:58:57.992 OurCrowd[512:139471] active
2016-04-25 11:58:57.993 OurCrowd[512:139471] PushPlugin skip clear badge
2016-04-25 11:59:01.181 OurCrowd[512:139471] uuid is: 
2016-04-25 11:59:01.181 OurCrowd[512:139471] ignore deploy
2016-04-25 11:59:01.181 OurCrowd[512:139471] ignore version: f7048708-0a19-11e6-8393-6a2cae197324


anyone has an idea what to do?

**Comment by Fuiste** [comment]
@asafco4 It's possible you'll have to remove the iOS platform then re-add it, as plugins can sometimes be cached.

@adinu when you say the app doesn't open, is it hanging on the splashscreen, or displaying some other behavior?

**Comment by adinu** [comment]
hanging on splashscreen.

**Comment by deliverymanager** [comment]
Could you please take a screen shot of the logs in xcode?

**Comment by deliverymanager** [comment]
Looks like this could be related to my issue I recently opened. It hangs on the same spot when opening the app.

**Comment by adinu** [comment]
this is the log we have on IOS:
2016-04-25 11:58:57.880 OurCrowd[512:139471] Apache Cordova native platform version 4.1.1 is starting.
2016-04-25 11:58:57.881 OurCrowd[512:139471] Multi-tasking -> Device: YES, App: YES
2016-04-25 11:58:57.963 OurCrowd[512:139471] Using UIWebView
2016-04-25 11:58:57.965 OurCrowd[512:139471] [CDVTimer][handleopenurl] 0.065982ms
2016-04-25 11:58:57.966 OurCrowd[512:139471] Unlimited access to network resources
2016-04-25 11:58:57.966 OurCrowd[512:139471] [CDVTimer][intentandnavigationfilter] 1.505971ms
2016-04-25 11:58:57.966 OurCrowd[512:139471] [CDVTimer][gesturehandler] 0.043988ms
2016-04-25 11:58:57.978 OurCrowd[512:139471] [CDVTimer][splashscreen] 11.386991ms
2016-04-25 11:58:57.985 OurCrowd[512:139471] [CDVTimer][statusbar] 6.640971ms
2016-04-25 11:58:57.985 OurCrowd[512:139471] [CDVTimer][keyboard] 0.503004ms
2016-04-25 11:58:57.986 OurCrowd[512:139471] VERSION LABEL: 1.4.6:1461574633:f7048708-0a19-11e6-8393-6a2cae197324
2016-04-25 11:58:57.986 OurCrowd[512:139471] updating version label
2016-04-25 11:58:57.987 OurCrowd[512:139471] [CDVTimer][ionicdeploy] 1.514018ms
2016-04-25 11:58:57.987 OurCrowd[512:139471] [CDVTimer][TotalPluginStartup] 22.298992ms
2016-04-25 11:58:57.991 OurCrowd[512:139471] createNotificationChecker
2016-04-25 11:58:57.991 OurCrowd[512:139471] not coldstart
2016-04-25 11:58:57.992 OurCrowd[512:139471] active
2016-04-25 11:58:57.993 OurCrowd[512:139471] PushPlugin skip clear badge
2016-04-25 11:59:01.181 OurCrowd[512:139471] uuid is: 
2016-04-25 11:59:01.181 OurCrowd[512:139471] ignore deploy
2016-04-25 11:59:01.181 OurCrowd[512:139471] ignore version: f7048708-0a19-11e6-8393-6a2cae197324

**Comment by martyzz1** [comment]
Same issue for me

0.5.0 Works fine...

0.5.2 also fails as above

Your system information:

Cordova CLI: 6.1.1
Gulp version:  CLI version 3.9.1
Gulp local:   Local version 3.9.1
Ionic Version: 1.2.4-nightly-1917
Ionic CLI Version: 1.7.14
Ionic App Lib Version: 0.7.0
ios-deploy version: 1.8.6
ios-sim version: 5.0.8
OS: Mac OS X El Capitan
Node Version: v4.4.2
Xcode version: Xcode 7.2.1 Build version 7C1002

output of failing logfile

2016-04-25 13:34:41.423 PawSquad[1363:663975] Apache Cordova native platform version 4.1.1 is starting.
2016-04-25 13:34:41.426 PawSquad[1363:663975] Multi-tasking -> Device: YES, App: YES
2016-04-25 13:34:41.722 PawSquad[1363:663975] Using UIWebView
2016-04-25 13:34:41.726 PawSquad[1363:663975] [CDVTimer][handleopenurl] 0.271976ms
2016-04-25 13:34:41.733 PawSquad[1363:663975] Unlimited access to network resources
2016-04-25 13:34:41.733 PawSquad[1363:663975] [CDVTimer][intentandnavigationfilter] 7.264018ms
2016-04-25 13:34:41.734 PawSquad[1363:663975] [CDVTimer][gesturehandler] 0.256002ms
2016-04-25 13:34:41.734 PawSquad[1363:663975] Starting Facebook Connect plugin
2016-04-25 13:34:41.735 PawSquad[1363:663975] [CDVTimer][facebookconnectplugin] 0.478983ms
2016-04-25 13:34:41.775 PawSquad[1363:663975] [CDVTimer][splashscreen] 39.902985ms
2016-04-25 13:34:41.776 PawSquad[1363:663975] VERSION LABEL: 1.1.5:1461587279:
2016-04-25 13:34:41.777 PawSquad[1363:663975] updating version label
2016-04-25 13:34:41.781 PawSquad[1363:663975] [CDVTimer][ionicdeploy] 5.468011ms
2016-04-25 13:34:41.982 PawSquad[1363:663975] [CDVTimer][statusbar] 201.026976ms
2016-04-25 13:34:41.983 PawSquad[1363:663975] [CDVTimer][keyboard] 0.569999ms
2016-04-25 13:34:41.983 PawSquad[1363:663975] [CDVTimer][googleplus] 0.185013ms
2016-04-25 13:34:41.983 PawSquad[1363:663975] [CDVTimer][TotalPluginStartup] 258.144021ms
2016-04-25 13:34:42.015 PawSquad[1363:663975] active
2016-04-25 13:34:42.015 PawSquad[1363:663975] PushPlugin skip clear badge
2016-04-25 13:34:42.628 PawSquad[1363:663975] uuid is: 
2016-04-25 13:34:42.628 PawSquad[1363:663975] ignore deploy
2016-04-25 13:34:42.629 PawSquad[1363:663975] ignore version:

**Comment by asafco4** [comment]
@Fuiste - I'm on Android right now, the plugin for ios and android is the same, right?
should be same version, right?

**Comment by Fuiste** [comment]
@asafco4 yup, should be the same versions, though version 0.5.1 made no changes to the Android side of things.

@martyzz1, which version of `cordova-ios` are you running?

**Comment by asafco4** [comment]
@Fuiste  - thanks for the quick reply.
I'm more concerned that I'm not getting the last version rather than the actual update, as right now I might not need to but later on I might...

I removed both platforms - ionic platform rm android & ionic platform rm ios
I then deleted the platforms & plugins folders from disk, then "ionic state restore"
(I can see the command: cordova plugin add ionic-plugin-deploy)
BUT I still see 0.5.0 :(

**Comment by Fuiste** [comment]
@asafco4 Probably a silly question, but have you tried simply running 

```
ionic plugin add ionic-plugin-deploy@0.5.2
```

I ask because I was just able to update the plugin on a local project in this manner.

@adinu I'm trying to reproduce this issue on my device now.  Just to pinpoint all variables, what version of iOS is your testing device running?

**Comment by asafco4** [comment]
@Fuiste  - I tried both with and without the @0.5.2
both installed the 0.5.0 

When I use ionic plugin add ionic-plugin-deploy - it should download the latest, right?
Could it be that since no change to Android then 0.5.0 is the latest downloadable?

Also, if I use specific version then it will be stuck on it while I want to get the latest...

**Comment by Fuiste** [comment]
Well, I feel foolish...

@asafco4 looks like I forgot to update the `plugin.xml` file when I published 0.5.2, so that one's on me.

I've just published 0.5.3, which is identical, save for the proper version bump and `plugin.xml` update, and should solve your issue.

**Comment by asafco4** [comment]
@Fuiste - Thanks :+1:

**Comment by martyzz1** [comment]
cordova platform version ios
Installed platforms:
  android 5.0.0
  ios 4.1.1

**Comment by martyzz1** [comment]
tried 0.5.3 and still get the issue...

**Comment by Fuiste** [comment]
Hey guys (@martyzz1, @adinu, and @deliverymanager), quick update.

It looks like I spoke too soon with that patch in 0.5.2, but I've just released v0.5.4 which is working on `cordova-ios` 3.9.2, 4.0.0, and 4.1.0 with both ym iOS 9.3 and 8.4 devices.

Let me know if this one works out for you, and sorry for the confusion there!
